### PR TITLE
feat(lanelet2_extension): add route checker

### PR DESCRIPTION
**Pull request by shmpwk**
## Description

Check the loading route is valid or not.
If it is invalid, puts warning without dying.
The case valid is when the route is created on the same map as the current one.

## Related links

This library will be used in the following PR.
https://github.com/autowarefoundation/autoware.universe/pull/1069
## Tests performed

I tested both valid and invalid route.

## Notes for reviewers

You can test this library by executing
```
colcon test --packages-select lanelet2_extension --ctest-args -R route-test
```

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: map/lanelet2_extension/include/lanelet2_extension/utility/route_checker.hpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#ifndef LANELET2_EXTENSION__UTILITY__ROUTE_CHECKER_HPP_
+#define LANELET2_EXTENSION__UTILITY__ROUTE_CHECKER_HPP_
+
+#include <lanelet2_extension/utility/query.hpp>
+#include <rclcpp/rclcpp.hpp>
+
+#include <autoware_auto_mapping_msgs/msg/had_map_bin.hpp>
+#include <autoware_auto_planning_msgs/msg/had_map_route.hpp>
+
+namespace lanelet
```
Comment (kenji-miyake) [inline map/lanelet2_extension/include/lanelet2_extension/utility/route_checker.hpp:24]: FYI: You can define nested namespaces like this from C++17.

```suggestion
namespace lanelet::utils::route
```

File: map/lanelet2_extension/test/src/test_route_checker.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+using lanelet::LineString3d;
+using lanelet::Point3d;
+using lanelet::utils::getId;
+
+class TestSuite : public ::testing::Test
+{
+public:
+  TestSuite() : sample_map_ptr(new lanelet::LaneletMap())
+  {
+    // NOLINT
```
Comment (kenji-miyake) [inline map/lanelet2_extension/test/src/test_route_checker.cpp:33]: Why NOLINT? :thinking:

File: map/lanelet2_extension/test/src/test_route_checker.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    // create sample lanelets
+    Point3d p1, p2, p3, p4;
+
+    p1 = Point3d(getId(), 0., 0., 0.);
+    p2 = Point3d(getId(), 0., 1., 0.);
+
+    LineString3d ls_left(getId(), {p1, p2});  // NOLINT
+
+    p3 = Point3d(getId(), 1., 0., 0.);
+    p4 = Point3d(getId(), 1., 1., 0.);
```
Comment (kenji-miyake) [inline map/lanelet2_extension/test/src/test_route_checker.cpp:43]: [IMO] This style is widely recommended.

```suggestion
    const Point3d p1(getId(), 0.0, 0.0, 0.0);
    const Point3d p2(getId(), 0.0, 1.0, 0.0);

    const LineString3d ls_left(getId(), {p1, p2});

    const Point3d p3(getId(), 1.0, 0.0, 0.0);
    const Point3d p4(getId(), 1.0, 0.0, 0.0);
```

File: map/lanelet2_extension/test/src/test_route_checker.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+};
+
+TEST_F(TestSuite, isRouteValid)
+{
+  autoware_auto_mapping_msgs::msg::HADMapBin bin_msg;
+
+  autoware_auto_planning_msgs::msg::HADMapRoute::ConstSharedPtr route_ptr1;
+  route_ptr1 = std::make_shared<autoware_auto_planning_msgs::msg::HADMapRoute>(sample_route1);
+  autoware_auto_planning_msgs::msg::HADMapRoute::ConstSharedPtr route_ptr2;
+  route_ptr2 = std::make_shared<autoware_auto_planning_msgs::msg::HADMapRoute>(sample_route2);
```
Comment (kenji-miyake) [inline map/lanelet2_extension/test/src/test_route_checker.cpp:81]: Same here. Changed to the widely recommended style.

```suggestion
  const auto route_ptr1 = std::make_shared<autoware_auto_planning_msgs::msg::HADMapRoute>(sample_route1);
  const auto route_ptr2 = std::make_shared<autoware_auto_planning_msgs::msg::HADMapRoute>(sample_route2);
```

**Review by kenji-miyake — COMMENTED** [review]
@shmpwk Added some reviews about coding styles.

File: map/lanelet2_extension/test/src/test_route_checker.cpp (hunk 3 above)
Comment (shmpwk) [reply map/lanelet2_extension/test/src/test_route_checker.cpp:43]: Thank you! I'll use `const`.

File: map/lanelet2_extension/test/src/test_route_checker.cpp (hunk 2 above)
Comment (kenji-miyake) [reply map/lanelet2_extension/test/src/test_route_checker.cpp:33]: Please remove as many NOLINT as possible.
If you have to keep some NOLINT, please write the reason.

Comment (shmpwk) [reply map/lanelet2_extension/test/src/test_route_checker.cpp]: Ok, rm NOLINT.
https://github.com/autowarefoundation/autoware.universe/pull/1149/commits/ec5cac09671784fe5d4b2dd03f25f1d67937d836

File: map/lanelet2_extension/include/lanelet2_extension/utility/route_checker.hpp (hunk 1 above)
Comment (shmpwk) [reply map/lanelet2_extension/include/lanelet2_extension/utility/route_checker.hpp]: I'm greate to hear that thank you.
https://github.com/autowarefoundation/autoware.universe/pull/1149/commits/ec5cac09671784fe5d4b2dd03f25f1d67937d836

**Review by tkimura4 — APPROVED** [review]
LGTM

**Comment by shmpwk** [comment]
Thank you. Then I will check if https://github.com/autowarefoundation/autoware.universe/pull/1153 works.

**Pull request merged**